src/java/org/apache/cassandra/db/ColumnFamilyStore.java

```diff
@@ -710,7 +713,20 @@ public static void removeUnfinishedCompactionLeftovers(CFMetaData metadata, Map<
             {
                 throw new FSReadError(e, "Failed to remove unfinished compaction leftovers (file: " + desc.filenameFor(Component.STATS) + ").  See log for details.");
             }
+            sstableToAncestors.put(desc, ancestors);
+        }
+
+        sstableToAncestors = ColumnFamilyStoreManager.instance.filterValidAncestors(metadata, sstableToAncestors, unfinishedCompactions);
+
+        for (Map.Entry<Descriptor, Set<Component>> sstableFiles : directories.sstableLister().skipTemporary(true).list().entrySet())
+        {
+            Descriptor desc = sstableFiles.getKey();
+            if (!sstableToAncestors.containsKey(desc))
```

Comment:
Instead of re-listing tables, why not loop over the `sstableToAncestors` result considering the `filterValidAncestors` already operates on the original `sstableToAncestors` map which is derived from `directories.sstableLister().skipTemporary(true).list().entrySet()` above?

Reply:
sure